xen/arch/x86/tpm.c

```diff
@@ -983,6 +976,181 @@ void __stdcall tpm_extend_mbi(uint32_t *mbi)
+    slrt = slr_get_table();
+
+    policy = slr_get_policy(slrt);
+    policy_entry = (struct slr_policy_entry *)
+        ((uint8_t *)policy + sizeof(*policy));
+
+    /* MBI policy entry must be the first one, so that measuring order matches
+     * policy order. */
+    if ( policy_entry[0].entity_type != SLR_ET_MULTIBOOT_INFO ) {
+        panic("First entry of DRTM policy in SLRT has wrong type: %#04x!\n",
+              policy_entry[0].entity_type);
+    }
+
+    verify_drtm_mbi_consistency(policy, policy_entry, mbi);
+
+    /*
+     * All SLR_POLICY_FLAG_MEASURED must be at the start of the list.
+     * MBI was measured in tpm_extend_mbi().
+     */
+    for ( i = 1; i < policy->nr_entries; i++ ) {
+        if ( !(policy_entry[i].flags & SLR_POLICY_FLAG_MEASURED) )
+            break;
+    }
+
+    for ( ; i < policy->nr_entries; i++ ) {
+        uint64_t start = policy_entry[i].entity;
+        uint64_t size = policy_entry[i].size;
+
+        if ( (policy_entry[i].flags & SLR_POLICY_FLAG_MEASURED) )
+            panic("DRTM entry at %d was measured out of order!\n", i);
+
+        /* Empty entity is considered invalid below, so this is to mark size as
+         * invalid. */
+        if ( policy_entry[i].flags & SLR_POLICY_IMPLICIT_SIZE )
+            size = 0;
+
+        switch ( policy_entry[i].entity_type ) {
+        case SLR_ET_MULTIBOOT_INFO:
+            panic("Duplicated MBI entry in DRTM of Secure Launch at %d\n", i);
+
+        case SLR_ET_SLRT:
+            if ( slrt->revision == 1 ) {
+                /*
+                 * In revision one of the SLRT, only Intel info table is
+                 * measured.  It's presence is checked by slr_get_policy().
+                 */
+                start = (uint64_t)slr_next_entry_by_tag(__va(start), NULL,
+                                                        SLR_ENTRY_INTEL_INFO);
+                size = sizeof(struct slr_entry_intel_info);
+            } else {
+                /*
+                 * slr_get_table() checks that revision is valid, so we must not
+                 * get here unless the code is wrong.
+                 */
+                panic("Unhandled SLRT revision: %d!\n", slrt->revision);
+            }
+            break;
+
+        case SLR_ET_UNSPECIFIED:
+        case SLR_ET_BOOT_PARAMS:
+        case SLR_ET_SETUP_DATA:
+        case SLR_ET_CMDLINE:
+        case SLR_ET_UEFI_MEMMAP:
+        case SLR_ET_RAMDISK:
+        case SLR_ET_MULTIBOOT_MODULE:
+        case SLR_ET_TXT_OS2MLE:
+            break;
+
+        case SLR_ET_UNUSED:
+            /* Skip this entry. */
+            continue;
+        }
+
+        if ( size == 0 )
+            panic("Can't measure DRTM policy entry at %d of size zero!\n", i);
+
+        map_pages_to_xen(start, maddr_to_mfn(start), PFN_UP(size),
+                         __PAGE_HYPERVISOR_RO);
+        tpm_hash_extend(DRTM_LOC, policy_entry[i].pcr, (uint8_t *)start, size,
+                        TXT_EVTYPE_SLAUNCH, NULL, 0);
```

Comment:
```suggestion
                        TXT_EVTYPE_SLAUNCH, policy_entry[i].evt_info, strlen(policy_entry[i].evt_info));
```

Maybe there is some more secure `strlen()` version, haven't checked.

Reply:
See https://github.com/TrenchBoot/xen/compare/c8f93f26c569ab37177393c2ceedb0cd012da4d1..cc6e675973ecc45d3e99d0c9b1301f16d3912476#diff-5861d423276bd98c31fb2a501569110297cbe6488842e0359435a3af795c24aeR1167-R1169

This however doesn't add description for `MBI` and `SLRT`.

Reply:
Let's leave those descriptions as possible future improvements. I understand that this implies depending on SLRT before it is measured, but in this case it isn't critical as it doesn't change the execution flow and TPM event log is purely informational.